### PR TITLE
S3 webhook feature data feed

### PR DESCRIPTION
**Pull request by shaik80**
### :nut_and_bolt: Description: What code changed, and why?

This PR  includes backend api's for data-feed service,
api is created for
1. Enable and disable for data-feed integration
2. Test connection 
	* Before saving data-feed integration
	* After saving data-feed integration, it also includes test connection by secret id(supports both old and new connection)
3. Add data-feed integration (it includes all integration)
4. Integration and unit test cases is included

### :chains: Related Resources
#5463 #5125
### :+1: Definition of Done
All test  cases should pass and api should be working as expected

### :athletic_shoe: How to Build and Test the Change
Goto automate folder enter 
```
1. hab studio enter
2. start_all_services
3. rebuild components/automate-gateway
4. rebuild components/data-feed-service/
5. hab pkg install --binlink core/curl -> if you don’t find  curl command in hab studio
```
how to run integration test cases
```
1. go inside hab studio
2. source integration/helpers/setup_minio.sh && start_minio_background -> ( setup minio in hab studio)
3. datafeed_run_integration_tests -> (to run integeration test cases)
```
how to run unit test cases
```
1. go automate repo
2. go test ./components/data-feed-service/service/*.go (to run unit test cases)
```
Make sure to check data-feed database
```
6. chef-automate dev psql postgres
7. \c data_feed_service
8. select * from destinations;
	id  | name | url  |  secret  | services | integration_types | meta_data  | enable
       -----+------+------+----------+----------+-------------------+------------+--------
```

Now we can run data-feed api's
- Add secrets for data-feed integration
 	*  request 
	 #### Storage (S3/Minio)
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/secrets' -d'{"name": "s3","type": "data_feed","data": [{"key":"access_key","value":"access_key_value"},{"key":"secret_access_key", "value":"secret_access_key_value"}]}'
	```
	#### Service Now 
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/secrets' -d'{"name": "serviceNOW","type": "data_feed","data":[{"key":"username","value":"admin"},{"key":"password","value":"Password"}]}'
	```

	#### Splunk
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/secrets' -d '{"name": "anyname","type": "data_feed","data": [{"key": "headers","value": "{\"Authorization\":\"Splunk 8a902a02-cc11-41f7-a2fa-e6a4297bd59d\"}"},{"key": "auth_type","value": "header_auth"}]}'
	```
	#### ELK and custom header 
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/secrets' -d '{"name": "anyname","type": "data_feed","data": [{"key": "headers","value": "{\"Authorization\":\"Bearer  <token>\"}"},{"key": "auth_type","value": "header_auth"}]}'
	```
	- 	 response
	```
	{"id":"38796b44-1821-4a8a-9002-4d9cd976871f"}
	```
-  Data-feed integration
	1.  Storage (S3) -> add
	- Request
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destination' -d '{"name": "S3","url":"null","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services": "S3","integration_types":"Storage","enable": true,"meta_data":[{"key":"bucket","value":"bucket_name"},{"key":"region","value":"us-east-2"}]}'
	```
	- Response
	```
	{"id":"285","name":"S3","url":"null","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services":"S3","integration_types":"Storage","enable":true,"meta_data":[{"key":"bucket","value":"bucket_name"},{"key":"region","value":"us-east-2"}]}
	```
	2.  Storage (Minio) -> add
	- Request
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destination' -d '{"name": "Minio","url":"http://127.0.0.1:9000","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services": "Minio","integration_types":"Storage","enable": true,"meta_data":[{"key":"bucket","value":"bucket_name"}]}'
	```
	-  Response
	```
	{"id":"286","name":"Minio","url":"http://127.0.0.1:9000","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services":"Minio","integration_types":"Storage","enable":true,"meta_data":[{"key":"bucket","value":"bucket_name"}]}
	```
	3.  Webhook (Service Now) -> add
	- Request
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destination' -d '{"name": "Service Now","url":"https://localhost/api/x_chef_automate/asset","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services": "Service Now","integration_types":"Webhook","enable": true,"meta_data":[]}'
	```
	-  Response
	```
	{"id":"286","name":"Service Now","url":"https://localhost/api/x_chef_automate/asset","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services":"Service Now","integration_types":"Webhook","enable":true,"meta_data":[]}
	```
	4.  Webhook (Splunk) -> add
	- Request
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destination' -d '{"name": "Splunk","url":"https://localhost","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services": "Splunk","integration_types":"Webhook","enable": true,"meta_data":[]}'
	```
	-  Response
	```
	{"id":"286","name":"Splunk","url":"https://localhost","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services":"Splunk","integration_types":"Webhook","enable":true,"meta_data":[]}
	```
	5.  Webhook (ELK/Custom) -> add
	- Request
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destination' -d '{"name": "Splunk","url":"https://localhost","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services": "ELK","integration_types":"Webhook","enable": true,"meta_data":[]}'
	```
	-  Response
	```
	{"id":"286","name":"ELK","url":"https://localhost","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services":"ELK","integration_types":"Webhook","enable":true,"meta_data":[]}
	```
	6. Enable and disable api
	- Request
	#### Enable
	```
	curl -sSkH "api-token: $(get_admin_token)" -X PATCH https://a2-dev.test/api/v0/datafeed/destination/enable/${id} -d'{ "id":"${id}","enable":true}'
	```
	#### Disable
	```
	curl -sSkH "api-token: $(get_admin_token)" -X PATCH https://a2-dev.test/api/v0/datafeed/destination/enable/${id} -d'{ "id":"${id}","enable":false}'
	```
	-  Response
	```
	{"id": "1972","name": "vendor","url": "http://127.0.0.1:9000","secret": "54d23262-232a-4c24-af70-f3b729586d3f","services": "Service Now","integration_types": "Webhook","enable": true,"meta_data": []}
	```
	7. Test connection api Before saving data-feed integration
	- Request
	#### Minio
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destinations/test' -d'{"url":"http://127.0.0.1:9000", "aws": {"access_key":"access_key_value","secret_access_key":"secret_access_key_value","bucket":"bucket_value","region":"us-west-2"}}'
	```
	#### S3
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destinations/test' -d'{"url":"null", "aws": {"access_key":"access_key_value","secret_access_key":"secret_access_key_value","bucket":"bucket_value","region":"us-west-2"}}'
	```

	#### ELK,costume headers and Splunk
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destinations/test' -d '{"url": "http://127.0.0.1:8088/services/collector/raw","header": { "value": "{\"Authorization\":\"Splunk 6f01b869-c181-4fb2-a74c-b619e6197a85\",\"bla\":\"dublu\"}"}}'
	```
	- Response
	```
	{"success": true}
	```
	8. Test connection api After saving data-feed integration
	#### Service Now/Splunk (To support old connection)
	- Request
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destinations/test' -d {"url": "https://localhost/api/x_chef_automate/asset","secret_id": {"id": "7ae6dc99-0518-42f6-9949-3617b6c436fb"}}
	```
	#### Service Now/Splunk  (To support new connection -> Webhook)
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destinations/test' -d { "url": "http://ec2-13-233-122-5.ap-south-1.compute.amazonaws.com:9000", "secret_id_with_addon": {"id": "e4299bf6-0d5d-490c-9352-c61ff707b455", "services": "Service Now", "integration_types": "Webhook","enable":true,"meta_data": []}}
	```
	#### S3 (To support new connection -> Storage)
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destinations/test' -d { "url": "null", "secret_id_with_addon": {"id": "e4299bf6-0d5d-490c-9352-c61ff707b455", "services": "S3", "integration_types": "Storage","enable":true,"meta_data": [{"key": "bucket", "value": "newtest"},{"key": "region", "value": "us-east-2"}]}}
	```
	#### Minio (To support new connection -> Storage)
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destinations/test' -d { "url": "http://127.0.0.1:9000", "secret_id_with_addon": {"id": "e4299bf6-0d5d-490c-9352-c61ff707b455", "services": "Minio", "integration_types": "Storage","enable":true,"meta_data": [{"key": "bucket", "value": "newtest"}]}}
	```
	- Response
	```
	{"success": true}	
	```
	9. Update data-feed integration
	- Request
	```
	curl -sSkH "api-token: $(get_admin_token)" -X PATCH 'https://a2-dev.test/api/v0/datafeed/destination/${id}' -d '{"id":286,"name": "Minio","url":"http://127.0.0.1:9000","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services": "Minio","integration_types":"Storage","enable": false,"meta_data":[{"key":"bucket","value":"bucket_name"}]}'
	```
	-  Response
	```
	{"id":"286","name":"Minio","url":"http://127.0.0.1:9000","secret":"38796b44-1821-4a8a-9002-4d9cd976871f","services":"Minio","integration_types":"Storage","enable":false,"meta_data":[{"key":"bucket","value":"bucket_name"}]}
	```
	10. List data-feed integration
	- Request
	```
	curl -sSkH "api-token: $(get_admin_token)" -X POST 'https://a2-dev.test/api/v0/datafeed/destinations' -d ''
	```
	-  Response
	```
	{"destinations": [{"id": "369","name": "S3","url": "null","secret": "ee9bdf41-dd6f-4f92-8f1a-15f46d5b0367","services": "S3","integration_types":"Storage","enable":true,"meta_data": [{"key": "region","value": "us-east-2"},{"key": "bucket","value": "s3.to.elastic.search"}]}]}
	```
	11. List data-feed config
	- Request
	```
	curl -sSkH "api-token: $(get_admin_token)" -X GET 'https://a2-dev.test/api/v0/datafeed/config' -d ''
	```
	-  Response
	```
	{"feed_interval": "0.016667","node_batch_size": "50","updated_nodes_only": true,"disable_cidr_filter": true,"cidr_filter": ["0.0.0.0/0"],"accepted_status_codes": [200,201,202,203,204]}
	```

### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [x] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

**Review by vivekshankar1 — COMMENTED** [review]
In List data-feed config response, how why do we get decimal values for feed interval, isn't there a way to get the actual values what we set in the config

File: e2e/package.json [hunk 1]
```diff
@@ -17,12 +17,10 @@
   },
   "author": "",
   "license": "ISC",
-  "dependencies": {
-    "cypress": "^4.1.0"
-  },
   "devDependencies": {
     "@cypress/webpack-preprocessor": "^4.1.0",
     "@types/js-base64": "^2.3.1",
+    "cypress": "^8.2.0",
```
Comment (vivekshankar1) [inline e2e/package.json:23]: Cypress tests are failing due to this change

File: components/data-feed-service/service/data-feed-service.go [hunk 2]
```diff
@@ -298,3 +337,55 @@ func getHostAttributes(attributesJson map[string]interface{}) (string, string, s
...
+		}
+	} else {
+		config = &aws.Config{
+			Region:      aws.String(cred.GetValues().AwsCreds.region),
+			Credentials: credentials.NewStaticCredentials(string(cred.GetValues().AwsCreds.accesskey), string(cred.GetValues().AwsCreds.secretAccessKey), ""),
+		}
+	}
+	sess, err := session.NewSession(config)
+	if err != nil {
+		panic(err)
```
Comment (vivekshankar1) [inline components/data-feed-service/service/data-feed-service.go:358]: don't panic here, handle the error

File: api/external/data_feed/data_feed.proto [hunk 3]
```diff
@@ -132,15 +197,37 @@ message UsernamePassword {
...
+}
 
 message URLValidationRequest {
   string url = 1;
   oneof credentials {
     UsernamePassword username_password = 2;
     SecretId secret_id = 3;
+	SecretIdWithExtraPrams secret_id_with_addon = 4;
+	Header header = 5;
+	AWS aws = 6;
```
Comment (vivekshankar1) [inline api/external/data_feed/data_feed.proto:227]: 4 spaces here please remove

**Review by vivekshankar1 — COMMENTED** [review]
deleting the package-loack.json from the original repository will cause version issues. Please don't delete it from the repo instead try removing that file from the commit

File: components/automate-gateway/handler/datafeed.go [hunk 4]
```diff
@@ -102,3 +102,30 @@ func (a *Datafeed) ListDestinations(ctx context.Context, in *data_feed.ListDesti
...
+		return a.client.EnableDestination(ctx, inDomain)
+	}
+	err := protobuf.CallDomainService(in, inDomain, f, out)
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+//config
```
Comment (vivek-yadav) [inline components/automate-gateway/handler/datafeed.go:119]: put relevant comments

**Comment by shaik80** [comment]
> Deleting the package-lock.json from the original repository will cause version issues. Please don't delete it from the repo instead try removing that file from the commit

have resolved the issue of package-lock.json

**Comment by shaik80** [comment]
> In List data-feed config response, how why do we get decimal values for feed interval, isn't there a way to get the actual values what we set in the config. Also please indent the proto files properly with 2 spaces instead of 4

I have done changes of instead in proto file and done changes of data-feed config response before I use to only sending hours in feed interval now I am sending whole data

```json
{
    "feed_interval": "1m0s",
    "node_batch_size": "50",
    "updated_nodes_only": true,
    "disable_cidr_filter": true,
    "cidr_filter": [
        "0.0.0.0/0"
    ],
    "accepted_status_codes": [
        200,
        201,
        202,
        203,
        204
    ]
}
```

**Review by vivekshankar1 — APPROVED** [review]
Looks good to me, working fine, Thanks shaik

**Comment by vivekshankar1** [comment]
> > In List data-feed config response, how why do we get decimal values for feed interval, isn't there a way to get the actual values what we set in the config. Also please indent the proto files properly with 2 spaces instead of 4
> 
> I have done changes of instead in proto file and done changes of data-feed config response before I use to only sending hours in feed interval now I am sending whole data
> 
> ```json
> {
>     "feed_interval": "1m0s",
>     "node_batch_size": "50",
>     "updated_nodes_only": true,
>     "disable_cidr_filter": true,
>     "cidr_filter": [
>         "0.0.0.0/0"
>     ],
>     "accepted_status_codes": [
>         200,
>         201,
>         202,
>         203,
>         204
>     ]
> }
> ```

Thanks shaik, this looks good

File: .studio/data-feed-service [hunk 5]
```diff
@@ -55,5 +55,8 @@ function datafeed_run_integration_tests() {
   chef-automate config patch components/data-feed-service/integration_test/patch.toml
   api_token=$(get_admin_token)
   export AUTOMATE_API_TOKEN="${api_token}"
+  export AWS_REGION="${AWS_REGION}"
+  export AWS_SECRET_ACCESS_KEY="${AWS_SECRET_ACCESS_KEY}"
+  export AWS_ACCESS_KEY_ID="${AWS_ACCESS_KEY_ID}"
   go_test -v github.com/chef/automate/components/data-feed-service/integration_test/...
 }
```
Comment (MadhviAtroliya) [inline .studio/data-feed-service:62]: Add EOL

**Review by MadhviAtroliya — APPROVED** [review]

**Review by shivangi-singh1 — APPROVED** [review]

File: .studio/data-feed-service (hunk 5 above)
Comment (shaik80) [reply .studio/data-feed-service:62]: Done

**Review by vivek-yadav — APPROVED** [review]

**Comment by vivek-yadav** [comment]
@shaik80 can you confirm this PR changes are working with Existing UI, and existing Data of old data.

**Pull request merged**